src/mlpack/methods/decision_tree/mad_gain.hpp

```diff
@@ -89,15 +89,37 @@ class MADGain
    * @param weights Weights associated to each value.
    */
   template<bool UseWeights, typename VecType, typename WeightVecType>
-  static double Evaluate(const VecType& values,
-                         const WeightVecType& weights)
+  double Evaluate(const VecType& values,
+                  const WeightVecType& weights)
   {
     // Corner case: if there are no elements, the impurity is zero.
     if (values.n_elem == 0)
       return 0.0;
 
     return Evaluate<UseWeights>(values, weights, 0, values.n_elem);
   }
+
+  /**
+   * Calculates the output value for each leaf node for prediction.
+   */
+  template<bool UseWeights, typename ResponsesType, typename WeightsType>
+  double OutputLeafValue(const ResponsesType& responses,
+                         const WeightsType& weights)
+  {
+    if (UseWeights)
+    {
+      double accWeights, weightedSum;
+      WeightedSum(responses, weights, 0, responses.n_elem, accWeights,
+          weightedSum);
+      return weightedSum / accWeights;
+    }
+    else
+    {
+      double sum;
+      Sum(responses, 0, responses.n_elem, sum);
+      return sum / responses.n_elem;
+    }
+  }
```

Comment:
Yes, I agree on making the documentation more comprehensive. But, I am not sure if it will be worth it to add another template parameter for it. It is true that this is the same only for regression tree FitnessFunctions but it will be literally different for each and every loss function in gradient boosting trees and it will unnecessarily complicate the code there.

If we really want to reduce code duplication, then we can have a base `RegressionGain` class and implement it there. But, at this moment, I think that we don't need that too either. Maybe we can deal with it in future.

Reply:
Agreed---let's avoid the extra template parameter for now then. :+1: